assets/js/data/collections/reducers.js

```diff
@@ -0,0 +1,127 @@
+/**
+ * Internal dependencies
+ */
+import { ACTION_TYPES as types } from './action-types';
+import { hasExisting } from './utils';
+
+const updateState = (
+	state = {},
+	namespace,
+	modelName,
+	ids,
+	queryString,
+	items,
+	replace = false
+) => {
+	if ( ! state[ namespace ] ) {
+		state = {
+			...state,
+			[ namespace ]: {
+				[ modelName ]: {
+					[ ids ]: {
+						[ queryString ]: items,
+					},
+				},
+			},
+		};
+	} else if ( ! state[ namespace ][ modelName ] ) {
+		state = {
+			...state,
+			[ namespace ]: {
+				...state[ namespace ],
+				[ modelName ]: {
+					[ ids ]: {
+						[ queryString ]: items,
+					},
+				},
+			},
+		};
+	} else if ( ! state[ namespace ][ modelName ][ ids ] ) {
+		state = {
+			...state,
+			[ namespace ]: {
+				...state[ namespace ],
+				[ modelName ]: {
+					...state[ namespace ][ modelName ],
+					[ ids ]: {
+						[ queryString ]: items,
+					},
+				},
+			},
+		};
+	} else {
+		state = replace
+			? {
+					...state,
+					[ namespace ]: {
+						...state[ namespace ],
+						[ modelName ]: {
+							...state[ namespace ][ modelName ],
+							[ ids ]: {
+								[ queryString ]: items,
+							},
+						},
+					},
+			  }
+			: {
+					...state,
+					[ namespace ]: {
+						...state[ namespace ],
+						[ modelName ]: {
+							...state[ namespace ][ modelName ],
+							[ ids ]: {
+								...state[ namespace ][ modelName ][ ids ],
+								[ queryString ]: items,
+							},
+						},
+					},
+			  };
+	}
+	return state;
```

Comment:
Probably could just use lodash merge here.  I was trying to avoid any other dependencies because of this getting used on the frontend, but I think it's unavoidable that we'll load lodash on the frontend because of dependency on `wp.element` via `wp.data`.  Part of optimization might be investigating tree-shaking/custom builds etc for the frontend so ya I'll just switch to lodash here.

Reply:
err, forgot `_.merge` (and `_.set`) mutate the object so we can't use here.  I had thought originally about using immutable.js here but again, trying to keep package size down.  Let's leave as is for now, we can look at abstracting when/if we add more reducers in future iterations. There's lot's of flexibility here with changing things in the future because it's internal (I'll create an issue after merging this..but also might introduce something in this pull).

Reply:
Fix coming in c193c8a79a6